### ISSUE TITLE
About the Florence_encoder

### ISSUE THREAD

**Issue by wxxhaoshuai** [JiuhaiChen/Florence-VL#4]
The return value of the vision_tower used here has three items, which seems different from that in the https://huggingface.co/microsoft/Florence-2-large. I want to know how I can use the vision_tower of florence2 as shown in your code.

**Comment by JiuhaiChen** [comment]
Hi, thanks for your interest, this is the vision encoder after the first pretranin stage: https://huggingface.co/jiuhai/florence-llama-vision-tower,  you can refer to https://huggingface.co/jiuhai/florence-llama-vision-tower/blob/e57962fa29a176139d2e22ffa3897af5075b55fa/modeling_florence2.py#L2857
for processing different text prompts, and this is the function  calling the vision encoder and get three different vision embeddings https://huggingface.co/jiuhai/florence-llama-vision-tower/blob/e57962fa29a176139d2e22ffa3897af5075b55fa/modeling_florence2.py#L2857

**Comment by wxxhaoshuai** [comment]
Thank you for your reply. I still want to know how to obtain the visual features [V′t1,V't2,V't3] that have interacted with the prompt in the paper.

**Comment by JiuhaiChen** [comment]
Hi, this is https://github.com/JiuhaiChen/Florence-VL/blob/main/llava/model/multimodal_encoder/florence_encoder.py#L47
for processing different text prompts, we tokenize the text prompt into input_id using https://huggingface.co/microsoft/Florence-2-large-ft/blob/main/tokenizer.json and feed the input_ids to the vision_tower to get different embedding. Let me know if you have further questions.

**Comment by wxxhaoshuai** [comment]
My understanding is that one task prompt corresponds to one visual feature. I want to know if it's possible to directly obtain the visual features of different tasks.It seems that the https://github.com/JiuhaiChen/Florence-VL/blob/c1d4abd2b60211af8c32a5202b7dc3f2a7454f09/llava/model/multimodal_encoder/florence_encoder.py#L62 only returns the visual features of DaVIT.

**Comment by JiuhaiChen** [comment]
Actually we modify the original Florence-2 generation code: https://huggingface.co/jiuhai/florence-llama-vision-tower/blob/e57962fa29a176139d2e22ffa3897af5075b55fa/modeling_florence2.py#L2857 
You can see for loop, each one process one prompt. 

https://github.com/JiuhaiChen/Florence-VL/blob/c1d4abd2b60211af8c32a5202b7dc3f2a7454f09/llava/model/multimodal_encoder/florence_encoder.py#L62 Return image features from different prompts (encoder_last_hidden_state) and DaVit (image_feature).

**Comment by wxxhaoshuai** [comment]
I see. Thank you for your answer

**Comment by wxxhaoshuai** [comment]
Sorry to bother you, but I was wondering if the florence2 visual encoder is trainable in both the pre-training phase and the fine-tuning phase? If so, in which code does he calculate the loss?

**Comment by JiuhaiChen** [comment]
Hi, we train the visual encoder in the pre-training phase but freeze it during the fine-tuning phase.  This is for the unfreeze the visual encoder: https://github.com/JiuhaiChen/Florence-VL/blob/main/llava/train/train.py#L1075

**Comment by wxxhaoshuai** [comment]
I find that freezing the visual encoder on trainable parameters is different from not freezing it on trainable parameters, but the amount of memory used for training is the same. Could you tell me why?

**Comment by JiuhaiChen** [comment]
Hi, I’m not sure how much memory can be saved by freezing the vision encoder, as the vision encoder is relatively small and still performs forward passes during training. I would expect the memory difference to be minor in this case.
